### ISSUE TITLE
Test-time Bernoulli sampling

### ISSUE THREAD

**Issue by steveazzolin** [SJTU-Quant/SNEX-GNN#1]
Dear authors,

I wonder how the test-time sampling from Bernoulli (as reported in *Explanation Generation* section, page 11751) is actually implemented.
Checking the code ([here](https://github.com/SJTU-Quant/SUNNY-GNN/blob/master/models/sunnygnn.py#L53)), I just see that during training a hard Gumbel distribution is used, while at test time a simple sigmoid is employed.
Hence, I do not understand where the Bernoulli sampling, and where the the number of sampled edges (according to parameter *k*) are implemented.

Thanks for the clarification,
Steve

**Comment by JialeDeng** [comment]
Dear Steve,

Thanks a lot for your attention to our work.
In the test time, the explainer outputs the importance scores for edges in the graph, and Bernoulli sampling can be performed based on the importance scores. Parameter k (num of sampled edges) should be set according to the specific sparsity requirements. You can find similar implementations in [GSAT](https://github.com/Graph-COM/GSAT/blob/6812f5ca6c4c93149cbd13c44e9f022d18861908/example/gsat.py#L64) and [PGExplainer](https://github.com/flyingdoog/PGExplainer/blob/51edc7b3000e980635f68618af027ae24297f6e5/codes/Explainer.py#L49). If you use dgl, you may use the [subgraph extraction ops](https://docs.dgl.ai/en/0.9.x/api/python/dgl.html#subgraph-extraction-ops) to construct a new explainable subgraph. If you use pyg, you may find [this code](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/explain/explanation.html#Explanation.get_explanation_subgraph) helpful.
Please let us know if you have following questions.

Best wishes,
Jiale

**Comment by steveazzolin** [comment]
Thanks for your answer, and thank you for pointing me to the references which are definitely helpful.
However, this does not fully address my question, as I don't find where this sampling is performed in your actual implementation.

Could you please point me to which part of the close this is actually applied?
Thank you again.

Cheers,
Steve
